src/Microsoft.AspNet.Hosting/Startup/FailSafeStartupLoader.cs

```diff
@@ -0,0 +1,162 @@
+            _runtimeEnv = runtimeEnv;
+            ShowDetailedErrors = hostingEnv.IsDevelopment()
+                || string.Equals("true", Environment.GetEnvironmentVariable("ASPNET_DETAILED_ERRORS"), StringComparison.OrdinalIgnoreCase);
+        }
+
+        // Internal for testing
+        internal bool ShowDetailedErrors { get; set; }
+
+        public Type FindStartupType(string startupAssemblyName, IList<string> diagnosticMessages)
+        {
+            try
+            {
+                var type = _nestedLoader.FindStartupType(startupAssemblyName, diagnosticMessages);
+                if (type == null)
+                {
+                    _startupException = new ArgumentException(
+                        diagnosticMessages.Aggregate("Failed to find a startup type for the web application.", (a, b) => a + "\r\n" + b),
+                        startupAssemblyName);
+                }
+                return type ?? typeof(FailSafeStartup);
+            }
+            catch (Exception ex)
+            {
+                _startupException = ex;
+                return typeof(FailSafeStartup);
+            }
+        }
+
+        public StartupMethods LoadMethods(Type startupType, IList<string> diagnosticMessages)
+        {
+            if (typeof(FailSafeStartup).Equals(startupType))
+            {
+                var errorText = _startupException?.ToString();
+                return new StartupMethods(new FailSafeStartup(ShowDetailedErrors, _runtimeEnv, _startupException).Configure);
+            }
+
+            try
+            {
+                var methods = _nestedLoader.LoadMethods(startupType, diagnosticMessages);
+                if (methods == null)
+                {
+                    var exception = new ArgumentException(
+                        diagnosticMessages.Aggregate("Failed to find a startup entry point for the web application.", (a, b) => a + "\r\n" + b),
+                        startupType.ToString());
+                    return new StartupMethods(new FailSafeStartup(ShowDetailedErrors, _runtimeEnv, exception).Configure);
+                }
+                var wrapper = new FailSafeStartupWrapper(ShowDetailedErrors, _runtimeEnv, methods.ConfigureServicesDelegate, methods.ConfigureDelegate);
+                return new StartupMethods(wrapper.Configure, wrapper.ConfigureServices);
+            }
+            catch (Exception ex)
+            {
+                return new StartupMethods(new FailSafeStartup(ShowDetailedErrors, _runtimeEnv, ex).Configure);
+            }
+        }
+
+        internal class FailSafeStartup
+        {
+            private readonly byte[] _errorBytes;
+
+            public FailSafeStartup(bool showDetails, IRuntimeEnvironment env, Exception ex)
+            {
+                // Write errors to standard out so they can be retrieved when not in development mode.
+                Console.Out.WriteLine("Application startup exception: " + ex.ToString());
```

Comment:
Don't assume that, we should just log. You can imagine hosting registering an EventLogProvider on windows so that critical exceptions like this show up there.